### ISSUE TITLE
Chrome sometimes not able to establish a CONNECT tunnel

### ISSUE THREAD

**Issue by myleshorton** [getlantern/lantern#5514]
It looks like this can happen sometimes when the remote server closes the connection. This could happen, for example, if there's an idle connection the server decides to close. This will then kick out to this chunk of code in `client/handler.go`:

```
	if req.Method == http.MethodConnect {
		// CONNECT requests are often used for HTTPS requests.
		log.Tracef("Intercepting CONNECT %s", req.URL)
		err := client.interceptCONNECT(ctx, resp, req)
		if err != nil {
			log.Error(op.FailIf(err))
		}
	} else {
```

It will basically hit that error condition, which presumably just closes the connection to the browser, which causes Chrome to mark Lantern as a bad proxy.

From Chrome:
```
193973: HTTP_STREAM_JOB
https://ssl.gstatic.com/
Start Time: 2016-12-21 09:07:32.060

t=694279 [st=    0] +HTTP_STREAM_JOB  [dt=12903]
                     --> alternative_service = "Uninitialized :0"
                     --> original_url = "https://ssl.gstatic.com/"
                     --> priority = "LOWEST"
                     --> source_dependency = 193970 (URL_REQUEST)
                     --> url = "https://ssl.gstatic.com/"
t=694279 [st=    0]   +PROXY_SERVICE  [dt=0]
t=694279 [st=    0]      PROXY_SERVICE_RESOLVED_PROXY_LIST
                         --> pac_string = "PROXY 127.0.0.1:18345;DIRECT"
t=694279 [st=    0]   -PROXY_SERVICE
t=694279 [st=    0]   +HOST_RESOLVER_IMPL_REQUEST  [dt=0]
                       --> address_family = 0
                       --> allow_cached_response = true
                       --> host = "ssl.gstatic.com:443"
                       --> is_speculative = false
t=694279 [st=    0]      HOST_RESOLVER_IMPL_IPV6_REACHABILITY_CHECK
                         --> cached = true
                         --> ipv6_available = true
t=694279 [st=    0]   -HOST_RESOLVER_IMPL_REQUEST
                       --> net_error = -804 (ERR_DNS_CACHE_MISS)
t=694279 [st=    0]   +SOCKET_POOL  [dt=12829]
t=707108 [st=12829]      SOCKET_POOL_BOUND_TO_CONNECT_JOB
                         --> source_dependency = 193790 (CONNECT_JOB)
t=707108 [st=12829]   -SOCKET_POOL
                       --> net_error = -111 (ERR_TUNNEL_CONNECTION_FAILED)
t=707108 [st=12829]    PROXY_LIST_FALLBACK
                       --> bad_proxy = "127.0.0.1:18345"
t=707108 [st=12829]   +HOST_RESOLVER_IMPL_REQUEST  [dt=0]
                       --> address_family = 0
                       --> allow_cached_response = true
                       --> host = "ssl.gstatic.com:443"
                       --> is_speculative = false
t=707108 [st=12829]      HOST_RESOLVER_IMPL_IPV6_REACHABILITY_CHECK
                         --> cached = false
                         --> ipv6_available = true
t=707108 [st=12829]   -HOST_RESOLVER_IMPL_REQUEST
                       --> net_error = -804 (ERR_DNS_CACHE_MISS)
t=707110 [st=12831]   +SOCKET_POOL  [dt=71]
t=707181 [st=12902]      SOCKET_POOL_BOUND_TO_CONNECT_JOB
                         --> source_dependency = 194418 (CONNECT_JOB)
t=707181 [st=12902]      SOCKET_POOL_BOUND_TO_SOCKET
                         --> source_dependency = 194423 (SOCKET)
t=707181 [st=12902]   -SOCKET_POOL
t=707181 [st=12902]    HTTP_STREAM_REQUEST_PROTO
                       --> proto = "h2"
t=707181 [st=12902]   +HOST_RESOLVER_IMPL_REQUEST  [dt=0]
                       --> address_family = 0
                       --> allow_cached_response = true
                       --> host = "ssl.gstatic.com:443"
                       --> is_speculative = false
t=707181 [st=12902]      HOST_RESOLVER_IMPL_IPV6_REACHABILITY_CHECK
                         --> cached = true
                         --> ipv6_available = true
t=707181 [st=12902]      HOST_RESOLVER_IMPL_CACHE_HIT
t=707181 [st=12902]   -HOST_RESOLVER_IMPL_REQUEST
t=707181 [st=12902]    HTTP2_SESSION_POOL_IMPORTED_SESSION_FROM_SOCKET
                       --> source_dependency = 194424 (HTTP2_SESSION)
t=707182 [st=12903]    HTTP_STREAM_JOB_BOUND_TO_REQUEST
                       --> source_dependency = 193970 (URL_REQUEST)
t=707182 [st=12903] -HTTP_STREAM_JOB
```

From Lantern logs:

```
Dec 21 17:07:33.853 - 56m7s ERROR flashlight.client: handler.go:39 Error piping data to upstream: read tcp 127.0.0.1:18345->127.0.0.1:55463: read: connection reset by peer [error=Error piping data to upstream: %v error_location=github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).connect (proxy_connect.go:119) error_op=read error_text=Error piping data to upstream: read tcp 127.0.0.1:18345->127.0.0.1:55463: read: connection reset by peer error_type=errors.Error local_addr=127.0.0.1:55463 network=tcp op=proxy origin=encrypted-tbn1.gstatic.com:443 origin_host=encrypted-tbn1.gstatic.com origin_port=443 remote_addr=127.0.0.1:18345 root_op=proxy user_agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.95 Safari/537.36]
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).connect (proxy_connect.go:119)
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).(github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.connect)-fm (proxy_connect.go:47)
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/client.(*Client).serveHTTPWithContext (handler.go:37)
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/client.(*Client).ServeHTTP (handler.go:22)
ERROR flashlight.client: handler.go:39   at net/http.serverHandler.ServeHTTP (server.go:2202)
ERROR flashlight.client: handler.go:39   at net/http.(*conn).serve (server.go:1579)
ERROR flashlight.client: handler.go:39   at runtime.goexit (asm_amd64.s:2086)
ERROR flashlight.client: handler.go:39 Caused by: read: connection reset by peer
```

**Comment by myleshorton** [comment]
One thing to check is why exactly the proxy closed the connection.

**Comment by myleshorton** [comment]
In theory the client's idle timers should always be shorter than the server's, if idle timing is the cause.

**Comment by myleshorton** [comment]
Of course looking at the log more closely, it looks like the browser is the one that closed the connection!

```
handler.go:39 Error piping data to upstream: read tcp 127.0.0.1:18345->127.0.0.1:55463
```

**Comment by myleshorton** [comment]
Hmnn -- realized I was slightly misreading the Chrome error log. This is the correct log from a new run on the chrome and then lantern sides, interestingly again with ssl.gstatic.com

```
89849: HTTP_STREAM_JOB
https://ssl.gstatic.com/
Start Time: 2016-12-21 14:28:53.884

t=1961804 [st=    0] +HTTP_STREAM_JOB  [dt=20002]
                      --> alternative_service = "Uninitialized :0"
                      --> original_url = "https://ssl.gstatic.com/"
                      --> priority = "LOWEST"
                      --> source_dependency = 89846 (URL_REQUEST)
                      --> url = "https://ssl.gstatic.com/"
t=1961804 [st=    0]   +PROXY_SERVICE  [dt=0]
t=1961804 [st=    0]      PROXY_SERVICE_RESOLVED_PROXY_LIST
                          --> pac_string = "PROXY 127.0.0.1:18345;DIRECT"
t=1961804 [st=    0]   -PROXY_SERVICE
t=1961804 [st=    0]   +HOST_RESOLVER_IMPL_REQUEST  [dt=1]
                        --> address_family = 0
                        --> allow_cached_response = true
                        --> host = "ssl.gstatic.com:443"
                        --> is_speculative = false
t=1961805 [st=    1]      HOST_RESOLVER_IMPL_IPV6_REACHABILITY_CHECK
                          --> cached = true
                          --> ipv6_available = true
t=1961805 [st=    1]   -HOST_RESOLVER_IMPL_REQUEST
                        --> net_error = -804 (ERR_DNS_CACHE_MISS)
t=1961805 [st=    1]   +SOCKET_POOL  [dt=19889]
t=1981694 [st=19890]      SOCKET_POOL_BOUND_TO_CONNECT_JOB
                          --> source_dependency = 89834 (CONNECT_JOB)
t=1981694 [st=19890]   -SOCKET_POOL
                        --> net_error = -111 (ERR_TUNNEL_CONNECTION_FAILED)
t=1981694 [st=19890]    PROXY_LIST_FALLBACK
                        --> bad_proxy = "127.0.0.1:18345"
t=1981694 [st=19890]   +HOST_RESOLVER_IMPL_REQUEST  [dt=1]
                        --> address_family = 0
                        --> allow_cached_response = true
                        --> host = "ssl.gstatic.com:443"
                        --> is_speculative = false
t=1981695 [st=19891]      HOST_RESOLVER_IMPL_IPV6_REACHABILITY_CHECK
                          --> cached = false
                          --> ipv6_available = true
t=1981695 [st=19891]   -HOST_RESOLVER_IMPL_REQUEST
                        --> net_error = -804 (ERR_DNS_CACHE_MISS)
t=1981695 [st=19891]   +SOCKET_POOL  [dt=110]
t=1981805 [st=20001]      SOCKET_POOL_BOUND_TO_CONNECT_JOB
                          --> source_dependency = 89961 (CONNECT_JOB)
t=1981805 [st=20001]      SOCKET_POOL_BOUND_TO_SOCKET
                          --> source_dependency = 89966 (SOCKET)
t=1981805 [st=20001]   -SOCKET_POOL
t=1981806 [st=20002]    HTTP_STREAM_REQUEST_PROTO
                        --> proto = "h2"
t=1981806 [st=20002]   +HOST_RESOLVER_IMPL_REQUEST  [dt=0]
                        --> address_family = 0
                        --> allow_cached_response = true
                        --> host = "ssl.gstatic.com:443"
                        --> is_speculative = false
t=1981806 [st=20002]      HOST_RESOLVER_IMPL_IPV6_REACHABILITY_CHECK
                          --> cached = true
                          --> ipv6_available = true
t=1981806 [st=20002]      HOST_RESOLVER_IMPL_CACHE_HIT
t=1981806 [st=20002]   -HOST_RESOLVER_IMPL_REQUEST
t=1981806 [st=20002]    HTTP2_SESSION_POOL_IMPORTED_SESSION_FROM_SOCKET
                        --> source_dependency = 89971 (HTTP2_SESSION)
t=1981806 [st=20002]    HTTP_STREAM_JOB_BOUND_TO_REQUEST
                        --> source_dependency = 89846 (URL_REQUEST)
t=1981806 [st=20002] -HTTP_STREAM_JOB
```

```
Dec 21 22:29:13.759 - 6m14s DEBUG proxy: proxy_connect.go:83 Unable to dial upstream: context deadline exceeded [error=Unable to dial upstream: %s error_location=github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).connect (proxy_connect.go:82) error_op= error_text=Unab
le to dial upstream: context deadline exceeded error_type=errors.Error op=proxy origin=ssl.gstatic.com:443 origin_host=ssl.gstatic.com origin_port=443 root_op=proxy user_agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.95 Safari/537.36]
DEBUG proxy: proxy_connect.go:83   at github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).connect (proxy_connect.go:82)
DEBUG proxy: proxy_connect.go:83   at github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).(github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.connect)-fm (proxy_connect.go:47)
DEBUG proxy: proxy_connect.go:83   at github.com/getlantern/flashlight/client.(*Client).serveHTTPWithContext (handler.go:37)
DEBUG proxy: proxy_connect.go:83   at github.com/getlantern/flashlight/client.(*Client).ServeHTTP (handler.go:22)
DEBUG proxy: proxy_connect.go:83   at net/http.serverHandler.ServeHTTP (server.go:2202)
DEBUG proxy: proxy_connect.go:83   at net/http.(*conn).serve (server.go:1579)
DEBUG proxy: proxy_connect.go:83   at runtime.goexit (asm_amd64.s:2086)
DEBUG proxy: proxy_connect.go:83 Caused by: context deadline exceeded
Dec 21 22:29:13.760 - 6m14s DEBUG proxy: proxy.go:32 Responding BadGateway: Unable to dial upstream: context deadline exceeded [op=proxy origin=ssl.gstatic.com:443 origin_host=ssl.gstatic.com origin_port=443 root_op=proxy user_agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML,
 like Gecko) Chrome/55.0.2883.95 Safari/537.36]
Dec 21 22:29:13.761 - 6m14s ERROR flashlight.client: handler.go:39 Unable to dial upstream: context deadline exceeded [error=Unable to dial upstream: %s error_location=github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).connect (proxy_connect.go:82) error_op= error_tex
t=Unable to dial upstream: context deadline exceeded error_type=errors.Error op=proxy origin=ssl.gstatic.com:443 origin_host=ssl.gstatic.com origin_port=443 root_op=proxy user_agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.95 Safari/537.36]
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).connect (proxy_connect.go:82)
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.(*connectInterceptor).(github.com/getlantern/flashlight/vendor/github.com/getlantern/proxy.connect)-fm (proxy_connect.go:47)
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/client.(*Client).serveHTTPWithContext (handler.go:37)
ERROR flashlight.client: handler.go:39   at github.com/getlantern/flashlight/client.(*Client).ServeHTTP (handler.go:22)
ERROR flashlight.client: handler.go:39   at net/http.serverHandler.ServeHTTP (server.go:2202)
ERROR flashlight.client: handler.go:39   at net/http.(*conn).serve (server.go:1579)
ERROR flashlight.client: handler.go:39   at runtime.goexit (asm_amd64.s:2086)
ERROR flashlight.client: handler.go:39 Caused by: context deadline exceeded
```

**Comment by myleshorton** [comment]
@joesis I feel like you might have a better sense of how to handle this. It looks like we are responding with Bad Gateway, but chrome is still marking Lantern as a bad proxy,

**Comment by myleshorton** [comment]
A few interesting data points @joesis:

1) turning on chrome debug logging doesn't seem to turn up much
2) there are few different places in the chrome source where ERR_TUNNEL_CONNECTION_FAILED pops up

https://cs.chromium.org/search/?q=ERR_TUNNEL_CONNECTION_FAILED&sq=package:chromium&type=cs

**Comment by myleshorton** [comment]
Actually it looks quite likely this is what we're hitting @joesis:

https://cs.chromium.org/chromium/src/net/http/http_proxy_client_socket.cc?rcl=0&l=507

Since it looks like that's where a Bad Gateway response would reach...

No not really sure if there's a solid way around this TBH...

**Comment by myleshorton** [comment]
Oh, and I've definitely confirmed that responding with Bad Gateway consistently gets us marked as a bad proxy, which is somewhat surprising actually.

**Comment by myleshorton** [comment]
So this is interesting @joesis. I mistakenly tried this thinking maybe the problem wasn't with the Bad Gateway but rather us not closing the connection after the Bad Gateway.

This, change, though, of course causes a panic because `upstream` is nil. Interestingly, though, it also made it so that Chrome is not marked as a bad proxy after Bad Gateway responses!

```
	upstream, err = ic.dial(ctx, "tcp", req.URL.Host)
	if err != nil {
		fullErr := errors.New("Unable to dial upstream: %s", err)
		log.Debug(fullErr)
		respondBadGateway(w, fullErr)
		closeUpstream = true
		return fullErr
	}
```

**Comment by myleshorton** [comment]
OK about to submit a PR @joesis

**Comment by myleshorton** [comment]
Darn this happened again with that change - reopening.

**Comment by myleshorton** [comment]
One question is why the proxies themselves seem to fail on certain requests @joesis.

**Comment by myleshorton** [comment]
Here's what I've been doing in case it's helpful:

In on shell:

```
$ ./lantern &> out.txt
```

In another:

```
tail -f out.txt | grep "Responding BadGateway"
```

**Comment by myleshorton** [comment]
So most of the errors I'm seeing are for bogus domains. Like `bits.wikimedia.org` isn't resolving to anything for me. That creates a 502 on the proxy and then in the client, with Lantern ultimately getting marked as a bad proxy.

Not sure what the solution to that is, as it seems like Bad Gateway is the correct response, but it's interesting.

**Comment by joesis** [comment]
@myleshorton I could reproduce what you saw by throttling at 10Kbit/s  and 2% packet drop rate, then browse for about 5 minutes. What interesting is I never encountered (or maybe just unnoticed) the issue while using the new binary in the recent days, sometimes under poor network condition.

![image](https://cloud.githubusercontent.com/assets/22091642/21416114/05f4eb18-c7c5-11e6-90da-8437cb455572.png)

**Comment by joesis** [comment]
> So most of the errors I'm seeing are for bogus domains. Like bits.wikimedia.org isn't resolving to anything for me...

I guess that's because Lantern usually works well enough if the site is reachable. How long it took Chrome to raise the `ERR_TUNNEL_CONNECTION_FAILED` error? Lantern should generate the same response as before when accessing unresolvable sites.

**Comment by myleshorton** [comment]
@oxtoacart and I chatted about this in more detail earlier, but it seems like in general Chrome considers an OK response to a CONNECT request to signify that the proxy itself is "OK" and not necessarily that the CONNECT has actually connected. According to that logic, any "Bad Gateway" or none OK response would indicate that *proxy* is not OK, and it gets marked as a bad proxy.

Accordingly changing Lantern to always preemptively return "OK" seems to significantly reduce the frequency of Chrome marking Lantern as a bad proxy. Sadly it still doesn't completely eliminate the problem, as I was able to eventually get Chrome to mark Lantern as a bad proxy with no corresponding request data:

![image](https://cloud.githubusercontent.com/assets/1143966/21436713/c77aadbc-c834-11e6-9381-6b247a8bf81d.png)
